blueman/plugins/applet/KillSwitch.py

```diff
@@ -91,6 +91,7 @@ def io_event(self, _, condition):
         if condition & GLib.IO_ERR or condition & GLib.IO_HUP:
             return False
 
+        assert self._fd is not None
```

Comment:
Why, if os.open in on_load fails we crash?

Reply:
Hm, right, we should probably just remove `_fd = None` from this class so that it's not optional. :thinking:

Reply:
```
iochan = GLib.IOChannel.new_file("/dev/rfkill", "r")
self._iom = GLib.io_add_watch(iochan, GLib.IO_IN | GLib.IO_ERR | GLib.IO_HUP, self.io_event)
```

The IOChannel documentation states that the file will be closed when all references to it are dropped. And then in on_unload when the watch is removed the last reference should dropped.

Weird GLib.io_watch_add accepted just a file descriptor, it must convert it to a IOChannel in the background.

Reply:
Fully working patch below, tested with my usb dongles.

```patch
diff --git a/blueman/plugins/applet/KillSwitch.py b/blueman/plugins/applet/KillSwitch.py
index c8b3e2c2..1052247a 100644
--- a/blueman/plugins/applet/KillSwitch.py
+++ b/blueman/plugins/applet/KillSwitch.py
@@ -51,7 +51,6 @@ class KillSwitch(AppletPlugin):
 
     _switches: Dict[int, Switch] = {}
     _iom = None
-    _fd = None
     _enabled = True
     _hardblocked = False
 
@@ -60,17 +59,17 @@ class KillSwitch(AppletPlugin):
         self._connman_watch_id = Gio.bus_watch_name(Gio.BusType.SYSTEM, "net.connman", Gio.BusNameWatcherFlags.NONE,
                                                     self._on_connman_appeared, self._on_connman_vanished)
 
-        self._fd = os.open('/dev/rfkill', os.O_RDONLY | os.O_NONBLOCK)
+        channel = GLib.IOChannel.new_file("/dev/rfkill", "r")
+        if channel is None:
+            raise ImportError('Failed to open rfkill device')
 
-        self._iom = GLib.io_add_watch(self._fd, GLib.IO_IN | GLib.IO_ERR | GLib.IO_HUP, self.io_event)
+        self._iom = GLib.io_add_watch(channel, GLib.IO_IN | GLib.IO_ERR | GLib.IO_HUP, self.io_event)
 
     def on_unload(self):
         Gio.bus_unwatch_name(self._connman_watch_id)
         self._connman_proxy = None
         if self._iom:
             GLib.source_remove(self._iom)
-        if self._fd:
-            os.close(self._fd)
 
     def _on_connman_appeared(self, connection, name, owner):
         logging.info("%s appeared" % name)
@@ -87,11 +86,11 @@ class KillSwitch(AppletPlugin):
         logging.info("%s vanished" % name)
         self._connman_proxy = None
 
-    def io_event(self, _, condition):
+    def io_event(self, channel, condition):
         if condition & GLib.IO_ERR or condition & GLib.IO_HUP:
             return False
 
-        data = os.read(self._fd, RFKILL_EVENT_SIZE_V1)
+        data = channel.read(RFKILL_EVENT_SIZE_V1)
         if len(data) != RFKILL_EVENT_SIZE_V1:
             logging.warning("Bad rfkill event size")
             return True
```